File: android/app/src/main/java/org/rimequad/ime/theme/ThemeModel.kt
package org.rimequad.ime.theme

/**
 * 主題的型別安全模型（docs/theme-format.md §8）。
 *
 * 所有顏色都是 ARGB Int；所有長度單位是 dp；所有文字尺寸單位是 sp（見 §4.3 / §4.4）。
 * 這些物件是不可變的，解析完成後可安全跨執行緒共用。
 */

enum class Appearance { LIGHT, DARK }

enum class Orientation { HORIZONTAL, VERTICAL }

enum class CommentPosition { AFTER, BELOW, HIDDEN }

enum class PageIndicatorKind { ARROWS, DOTS, TEXT, NONE }

enum class ScrollMode { NONE, HORIZONTAL, EXPANDABLE }

enum class Placement { BELOW, ABOVE, AUTO }

enum class Backdrop { NONE, BLUR, VIBRANCY }

enum class HintPosition { TOP_RIGHT, TOP_CENTER, TOP_LEFT, BOTTOM_RIGHT, NONE }

enum class MotionCurve { STANDARD, LINEAR, DECELERATE, ACCELERATE }

enum class HapticStrength { NONE, LIGHT, MEDIUM, HEAVY }

data class FontStack(
    val family: List<String>,
    val weight: Int,
    val italic: Boolean,
    val scriptFallback: Map<String, List<String>>
) {
    /**
     * §8.4.2 的最終字體解析順序：script 分支 ++ 基礎清單 ++ 平台預設。
     * 重複項只保留首次出現。
     */
    fun resolveFor(script: String?, platformDefaults: List<String>): List<String> {
        val out = ArrayList<String>()
        if (script != null) scriptFallback[script]?.let { out.addAll(it) }
        out.addAll(family)
        out.addAll(platformDefaults)
        val seen = LinkedHashSet<String>()
        for (f in out) seen.add(f)
        return seen.toList()
    }
}

data class FontAsset(
    val family: String,
    val file: String,
    val weight: Int,
    val italic: Boolean
)

data class Typography(
    val respectSystemFontScale: Boolean,
    val fontScaleMin: Float,
    val fontScaleMax: Float,
    val fonts: Map<String, FontStack>,
    val assets: List<FontAsset>
) {
    fun font(name: String): FontStack = fonts[name] ?: ThemeDefaults.FONT_STACK

    /** §4.4 的有效字體縮放。 */
    fun effectiveScale(systemFontScale: Float): Float =
        if (!respectSystemFontScale) 1.0f
        else clampFloat(systemFontScale, fontScaleMin, fontScaleMax)
}

data class Metrics(
    val cornerRadius: Float,
    val borderWidth: Float,
    val padding: Float,
    val spacing: Float,
    val elevation: Float
)

data class LabelStyle(
    val show: Boolean,
    val format: String,
    val size: Float,
    val color: Int,
    val highlightColor: Int
) {
    /** §8.6.1：未知佔位符必須原樣保留。 */
    fun render(label: String, indexOnPage: Int): String =
        format.replace("{label}", label)
            .replace("{index}", (indexOnPage + 1).toString())
            .replace("{index0}", indexOnPage.toString())
}

data class TextStyle(val size: Float, val color: Int, val highlightColor: Int)

data class CommentStyle(
    val show: Boolean,
    val position: CommentPosition,
    val size: Float,
    val color: Int,
    val highlightColor: Int
)

data class ItemStyle(
    val paddingH: Float,
    val paddingV: Float,
    val spacing: Float,
    val cornerRadius: Float,
    val minWidth: Float,
    val background: Int,
    val highlightBackground: Int,
    val borderWidth: Float,
    val borderColor: Int,
    val highlightBorderWidth: Float,
    val highlightBorderColor: Int
)

data class SeparatorStyle(val show: Boolean, val color: Int, val width: Float)

data class PageIndicatorStyle(
    val show: Boolean,
    val kind: PageIndicatorKind,
    val color: Int,
    val disabledColor: Int,
    val size: Float
)

/** 候選呈現的共用部分。`bar` 與 `window` 各自持有一份已套用覆寫的副本。 */
data class CandidateStyle(
    val orientation: Orientation,
    val label: LabelStyle,
    val text: TextStyle,
    val comment: CommentStyle,
    val item: ItemStyle,
    val separator: SeparatorStyle,
    val pageIndicator: PageIndicatorStyle
)

data class ExpandButton(val show: Boolean, val color: Int, val size: Float)

/**
 * 工具列項目（§8.6.6.1）。
 *
 * 刻意重用 [LabelSource] 與 [KeyAction]：工具列項目就是「沒有 `send` 的鍵」，
 * 不該為它發明第二套詞彙。鍵面文字的解析順序、圖示退化、active 觸發條件
 * 都與按鍵完全相同（§9.6、§8.8.1），渲染器可以共用同一段程式碼。
 */
data class ToolbarItem(
    val icon: String?,
    val label: String,
    val labelFrom: LabelSource,
    val tap: KeyAction
)

/**
 * 無候選時候選列顯示的工具列。
 *
 * 這不是裝飾:它是方案切換的唯一入口,所以 [items] 必定含有
 * `schema:picker` 與 `settings`(解析器會補回被主題刪掉的必備項)。
 */
data class Toolbar(val show: Boolean, val items: List<ToolbarItem>) {
    fun hasAction(verb: ActionVerb): Boolean = items.any { it.tap.verb == verb }
}

/** 行動端候選列。桌面端不消費。 */
data class CandidateBar(
    val style: CandidateStyle,
    val height: Float,
    val background: Int,
    val borderTopWidth: Float,
    val borderTopColor: Int,
    val maxVisible: Int,
    val scroll: ScrollMode,
    val expandButton: ExpandButton,
    val showPreeditInline: Boolean,
    val emptyShowsToolbar: Boolean,
    val toolbar: Toolbar
)

data class Shadow(
    val show: Boolean,
    val radius: Float,
    val offsetX: Float,
    val offsetY: Float,
    val color: Int
)

/** 桌面端懸浮候選窗。行動端不消費。 */
data class CandidateWindow(
    val style: CandidateStyle,
    val background: Int,
    val cornerRadius: Float,
    val padding: Float,
    val borderWidth: Float,
    val borderColor: Int,
    val minWidth: Float,
    val maxWidth: Float,
    val placement: Placement,
    val offsetX: Float,
    val offsetY: Float,
    val followCaret: Boolean,
    val backdrop: Backdrop,
    val opacity: Float,
    val shadow: Shadow
)

data class Candidates(
    val shared: CandidateStyle,
    val bar: CandidateBar,
    val window: CandidateWindow
)

data class Caret(val show: Boolean, val color: Int, val width: Float, val blink: Boolean)

data class SelectionStyle(val color: Int, val textColor: Int)

data class Preedit(
    val show: Boolean,
    val size: Float,
    val color: Int,
    val background: Int,
    val paddingH: Float,
    val paddingV: Float,
    val cornerRadius: Float,
    val caret: Caret,
    val selection: SelectionStyle
)

/**
 * §8.8.0 的高度模型。
 *
 * **因果方向是 鍵寬 → 鍵高 → 鍵盤高**，而不是初稿的 螢幕高 → 鍵盤高 → 鍵高。
 * 初稿的方向讓鍵高綁在螢幕高度、鍵寬綁在螢幕寬度，長寬比因此完全失控：
 * 在 20:9 的長螢幕上鍵會被拉成 1:1.94，使用者看到的就是「被拉伸」。
 */
data class KeyGeometry(
    val aspect: Float,
    val keyHeightMin: Float,
    val keyHeightMax: Float,
    val maxScreenRatioPortrait: Float,
    val maxScreenRatioLandscape: Float
) {
    /** 一次算完的結果，供渲染器直接使用。 */
    data class Resolved(val keyWidth: Float, val keyHeight: Float, val keyboardHeight: Float)

    /**
     * @param widthDp     鍵盤可用寬度（通常是螢幕寬）
     * @param availHeightDp 當前方向下宿主視窗可用高度，只用於安全網
     * @param units       當前 layer 的欄數（§9.3）
     * @param rowsWeight  當前 layer 的 Σ row.weight（不是列數 —— 列可以有不同權重）
     * @param rowCount    當前 layer 的列數，只用來算列間距總和
     * @param heightScale 佈局的 `metrics.height_scale`（§9.2）
     */
    fun resolve(
        widthDp: Float,
        availHeightDp: Float,
        landscape: Boolean,
        units: Float,
        rowsWeight: Float,
        rowCount: Int,
        padding: EdgeInsets,
        keySpacing: Float,
        rowSpacing: Float,
        heightScale: Float = 1.0f
    ): Resolved {
        val safeUnits = if (units <= 0f) 1f else units
        val safeRowsW = if (rowsWeight <= 0f) 1f else rowsWeight
        val gaps = if (rowCount > 1) rowCount - 1 else 0

        // 1. 鍵寬：與螢幕高度無關
        val innerW = widthDp - padding.left - padding.right - keySpacing * (safeUnits - 1f)
        val keyW = (if (innerW > 0f) innerW else 0f) / safeUnits

        // 2. 鍵高：由鍵寬推導，夾制後才套 height_scale
        var keyH = clampFloat(keyW * aspect, keyHeightMin, keyHeightMax) * heightScale

        // 3. 鍵盤高是算出來的
        val chrome = rowSpacing * gaps + padding.top + padding.bottom
        var h = keyH * safeRowsW + chrome

        // 4. 安全網：不得超過螢幕的這個比例
        val ratio = if (landscape) maxScreenRatioLandscape else maxScreenRatioPortrait
        val cap = availHeightDp * ratio
        if (availHeightDp > 0f && h > cap) {
            keyH = ((cap - chrome) / safeRowsW).coerceAtLeast(1f)
            h = keyH * safeRowsW + chrome
        }
        return Resolved(keyWidth = keyW, keyHeight = keyH, keyboardHeight = h)
    }
}

data class EdgeInsets(val left: Float, val top: Float, val right: Float, val bottom: Float)

data class KeyStyle(
    val background: Int,
    val pressedBackground: Int,
    val foreground: Int,
    val pressedForeground: Int,
    val activeBackground: Int,
    val activeForeground: Int,
    val cornerRadius: Float,
    val borderWidth: Float,
    val borderColor: Int,
    val elevation: Float,
    val font: String,
    val labelSize: Float,
    val hintSize: Float,
    val hintColor: Int,
    val hintPosition: HintPosition,
    val iconSize: Float
)

data class PopupStyle(
    val show: Boolean,
    val background: Int,
    val foreground: Int,
    val highlightBackground: Int,
    val highlightForeground: Int,
    val cornerRadius: Float,
    val itemSize: Float,
    val itemPadding: Float,
    val elevation: Float,
    val maxColumns: Int
)

data class PressPreviewStyle(
    val show: Boolean,
    val background: Int,
    val foreground: Int,
    val size: Float,
    val cornerRadius: Float,
    val elevation: Float
)

/** 僅行動端消費；桌面端實作必須整段忽略。 */
data class Keyboard(
    val background: Int,
    val geometry: KeyGeometry,
    val padding: EdgeInsets,
    val rowSpacing: Float,
    val keySpacing: Float,
    val honorBottomInset: Boolean,
    val keyStyles: Map<String, KeyStyle>,
    val popup: PopupStyle,
    val pressPreview: PressPreviewStyle
) {
    /** §6.3：佈局引用不存在的 style 名 → 退回 `default`。 */
    fun keyStyle(name: String): KeyStyle =
        keyStyles[name] ?: keyStyles["default"] ?: ThemeDefaults.KEY_STYLE
}

data class Motion(
    val enabled: Boolean,
    val respectReduceMotion: Boolean,
    val curve: MotionCurve,
    val keyPressMs: Int,
    val keyReleaseMs: Int,
    val candidateChangeMs: Int,
    val popupMs: Int,
    val windowShowMs: Int
) {
    /** §8.9：系統開啟「減少動態效果」時所有 duration 視為 0。 */
    fun durationOf(base: Int, reduceMotionOn: Boolean): Int =
        if (!enabled || (respectReduceMotion && reduceMotionOn)) 0 else base
}

/** 僅行動端。分類上其實是「行為」而非「外觀」，見規範 §8.10 的說明。 */
data class Feedback(
    val haptic: Boolean,
    val hapticStrength: HapticStrength,
    val sound: Boolean,
    val soundVolume: Float
)

data class Theme(
    val id: String,
    val revision: Int,
    val name: LocalizedString,
    val description: LocalizedString,
    val author: String,
    val license: String,
    val appearance: Appearance,
    val counterpart: String?,
    /** 由根祖先到自身的 id 鏈，供除錯與主題編輯器使用。 */
    val ancestry: List<String>,
    val palette: Map<String, Int>,
    val typography: Typography,
    val metrics: Metrics,
    val candidates: Candidates,
    val preedit: Preedit,
    val keyboard: Keyboard,
    val motion: Motion,
    val feedback: Feedback
)

File: apple/LuminaKey/Sources/LuminaKeyKit/ThemeModel.swift
//
//  ThemeModel.swift — 主題的結果物件（docs/theme-format.md §8）
//
//  ⚠ 桌面端**不**建立 `keyboard` 與 `feedback` 的模型（§1.1：桌面端整段忽略），
//    也完全不碰 core/layouts/。這不是偷懶，是規範裡結構性的分野。
//    「不消費」不等於「可以刪」—— 使用者的行動端設定必須原樣搬運。
//
//  ⚠ `candidates.syllables`（§8.6.6.3）是**例外**：桌面端建立它的模型、
//    但不渲染它。理由見 `SyllableBar` 的註解 —— 那不是「準備要畫」。
//

import Foundation

public enum Appearance: String, Sendable { case light, dark }
public enum Orientation: String, Sendable { case horizontal, vertical }
public enum CommentPosition: String, Sendable { case after, below, hidden }
public enum PageIndicatorKind: String, Sendable { case arrows, dots, text, none }
public enum Placement: String, Sendable { case below, above, auto }
public enum Backdrop: String, Sendable { case none, blur, vibrancy }
/// §8.6.7.1（本輪新增）：格內對齊。
public enum ItemAlign: String, Sendable { case leading, center, trailing }
/// §8.6.7.1（本輪新增）：超出 `max_width` 時的處置。
public enum Overflow: String, Sendable { case shrink, clip }
/// §8.12（本輪新增）：狀態列的排列。
public enum Arrangement: String, Sendable { case leading, center, trailing, spaceBetween = "space_between" }
public enum StatusPosition: String, Sendable { case top, bottom }
/// §8.13（本輪新增）：候選朗讀的詳細程度。
public enum AnnounceLevel: String, Sendable { case full, textOnly = "text_only", none }

public struct Metrics: Sendable {
    public var cornerRadius = 8.0
    public var borderWidth = 0.0
    public var padding = 6.0
    public var spacing = 4.0
    public var elevation = 2.0
}

public struct FontStack: Sendable {
    public var family: [String] = ["$system"]
    public var weight = 400
    public var italic = false
    public var scriptFallback: [String: [String]] = [:]
}

public struct FontAsset: Sendable {
    public var family: String
    public var file: String
    public var weight: Int
    public var italic: Bool
}

public struct Typography: Sendable {
    public var respectSystemFontScale = true
    public var fontScaleMin = 0.85
    public var fontScaleMax = 1.30
    public var fonts: [String: FontStack] = [:]
    public var fontOrder: [String] = []
    public var assets: [FontAsset] = []

    /// §4.4：macOS 的 `system_font_scale` 固定為 1.0（系統無等價設定），
    /// 所以 `effective_scale` 恆為 1.0，而且 AppKit 不會自己再套一次。
    /// 這個函式存在只是為了讓那條規則在程式碼裡有一個名字，不是散落的魔術數。
    public func effectiveScale(systemFontScale: Double = 1.0) -> Double {
        respectSystemFontScale ? clamp(systemFontScale, fontScaleMin, fontScaleMax) : 1.0
    }
}

public struct LabelStyle: Sendable {
    public var show = true
    public var format = "{label}"
    public var size = 12.0
    public var color = RGBA.hex(0x808080FF)
    public var highlightColor = RGBA.hex(0x808080FF)
}

public struct TextStyle: Sendable {
    public var size = 20.0
    public var color = RGBA.hex(0x000000FF)
    public var highlightColor = RGBA.hex(0x000000FF)
}

public struct CommentStyle: Sendable {
    public var show = true
    public var position = CommentPosition.after
    public var size = 12.0
    public var color = RGBA.hex(0x808080FF)
    public var highlightColor = RGBA.hex(0x808080FF)
}

public struct ItemStyle: Sendable {
    public var paddingH = 6.0
    public var paddingV = 6.0
    public var spacing = 4.0
    public var cornerRadius = 8.0
    public var minWidth = 0.0
    public var background = RGBA.transparent
    public var highlightBackground = RGBA.hex(0x3060C0FF)
    public var borderWidth = 0.0
    public var borderColor = RGBA.transparent
    public var highlightBorderWidth = 0.0
    public var highlightBorderColor = RGBA.transparent
}

public struct SeparatorStyle: Sendable {
    public var show = false
    public var color = RGBA.hex(0x808080FF)
    public var width = 1.0
}

public struct PageIndicatorStyle: Sendable {
    public var show = true
    public var kind = PageIndicatorKind.arrows
    public var color = RGBA.hex(0x808080FF)
    public var disabledColor = RGBA.hex(0x808080FF)
    public var size = 14.0
}

/// §8.6.1–8.6.5：`candidates` / `candidates.bar` / `candidates.window` 共用的部分。
public struct CandidateStyle: Sendable {
    public var orientation = Orientation.horizontal
    public var label = LabelStyle()
    public var text = TextStyle()
    public var comment = CommentStyle()
    public var item = ItemStyle()
    public var separator = SeparatorStyle()
    public var pageIndicator = PageIndicatorStyle()

    public static func defaults(_ m: Metrics) -> CandidateStyle {
        var s = CandidateStyle()
        s.item.paddingH = m.padding
        s.item.paddingV = m.padding
        s.item.spacing = m.spacing
        s.item.cornerRadius = m.cornerRadius
        return s
    }
}

public struct Shadow: Sendable {
    public var show = true
    public var radius = 18.0
    public var offsetX = 0.0
    public var offsetY = 4.0
    public var color = RGBA.hex(0x00000040)
}

/// §8.6.7 桌面候選窗。`lines` 之後的欄位是本輪新增的 §8.6.7.1。
public struct CandidateWindow: Sendable {
    public var style = CandidateStyle()
    public var background = RGBA.hex(0xFFFFFFFF)
    public var cornerRadius = 8.0
    public var padding = 6.0
    public var borderWidth = 0.0
    public var borderColor = RGBA.transparent
    public var minWidth = 0.0
    public var maxWidth = 640.0
    public var placement = Placement.auto
    public var offsetX = 0.0
    public var offsetY = 6.0
    public var followCaret = true
    public var backdrop = Backdrop.none
    public var opacity = 1.0
    public var shadow = Shadow()

    // ── §8.6.7.1 多行／表格排版 ────────────────────────────────
    /// 次要軸上的行數。`1` = v1 的單行行為；`0` = 自動（在 `max_width` 內塞得下多少排多少）。
    public var lines = 1
    public var equalColumns = true
    public var columnGap = 4.0
    public var rowGap = 4.0
    public var maxHeight = 0.0
    public var itemAlign = ItemAlign.leading
    public var overflow = Overflow.shrink
}

/// §8.6.6.3：逐音節消歧欄畫在哪。
public enum SyllablePlacement: String, Sendable {
    case none
    case aboveCandidates = "above_candidates"
    case keyboardSlot = "keyboard_slot"
}

/// §8.6.6.3：什麼時候出現。
/// ⚠ `onDemand` **尚未有任何一端實作**（§8.6.6.3.4）—— 解析得出來不等於畫得出來。
public enum SyllableTrigger: String, Sendable {
    case whileComposing = "while_composing"
    case onDemand = "on_demand"
}

/// §8.6.6.3 逐音節消歧欄（九宮格上「MG 是 mi 還是 ni」那一欄）。
///
/// ⚠ **桌面端解析它、但不渲染它**（§8.6.6.3.5）。這個型別存在於桌面端的模型裡，
///   **不是**因為「快要畫了」，而是因為 §10 第 9 條要求四端對同一份壞主題報出
///   同一串診斷：桌面端若像 `candidates.bar` 那樣把整個區塊當成「已知但不進入」，
///   `candidates.syllables.placemnt` 這種拼字錯誤在手機上是一則、在電腦上是零則。
///
/// 桌面端沒有軟鍵盤，所以 `keyboardSlot` 在這一端是**形態上不存在**
///（同 §9.5.1 的 `hide_keyboard`）；`none` 與 `aboveCandidates` 不是 —— 候選窗
/// 完全有地方畫一橫排讀音，只是 v1 沒做。真的要做時形狀已經寫死在 §8.6.6.3.5，
/// 不要另外發明；擋在前面的也不是 UI 而是 ABI（`rs_snapshot` 只給一頁候選）。
public struct SyllableBar: Sendable {
    public var placement = SyllablePlacement.keyboardSlot
    public var trigger = SyllableTrigger.whileComposing
    /// 一次最多列幾個讀音；`0` = 有多少畫多少。
    public var maxItems = 0
    /// `above_candidates` 那一排的高度（dp）。`keyboard_slot` 不佔額外高度。
    public var height = 40.0
}

/// §8.12（本輪新增）狀態列項目的來源。
public enum StatusSource: String, Sendable {
    case schemaName = "schema_name"
    case schemaId = "schema_id"
    case inputMode = "input_mode"
    case inputModePair = "input_mode_pair"
    case shape
    case variant
    case page
    case text
}

public struct StatusItem: Sendable {
    public var source: StatusSource
    public var text: String = ""
    public var tap: KeyAction?

    public init(source: StatusSource, text: String = "", tap: KeyAction? = nil) {
        self.source = source
        self.text = text
        self.tap = tap
    }
}

/// §8.12（本輪新增）桌面候選窗的狀態列。
public struct StatusBar: Sendable {
    public var show = false
    public var position = StatusPosition.bottom
    public var background = RGBA.transparent
    public var height = 0.0
    public var paddingH = 6.0
    public var paddingV = 2.0
    public var spacing = 4.0
    public var arrangement = Arrangement.leading
    public var size = 11.0
    public var color = RGBA.hex(0x808080FF)
    public var activeColor = RGBA.hex(0x808080FF)
    public var separator = SeparatorStyle()
    public var items: [StatusItem] = StatusBar.defaultItems

    /// §8.12 的規範性預設清單。`items` 缺席時必須產生這一份。
    ///
    /// ⚠ **輸入模式那一項用 `input_mode` 而不是 `input_mode_pair`，而且是刻意的。**
    ///   `input_mode_pair`（同時畫「中」與「En」、強調當前那一態）存在的理由是
    ///   **按鍵**的歧義：一顆只寫「中」的鍵有兩種讀法（「現在是中文」與
    ///   「按了會變中文」），而它們指向相反的操作。
    ///   桌面的狀態列不是按鍵 —— 它是**顯示**（`CandidateView` 只對候選格
    ///   做 hit-test，狀態列點不下去），所以那個歧義不存在，而兩態並排反而
    ///   變成「兩個都畫出來讓使用者猜現在是哪一個」。真機回報過同一件事。
    ///   行動端的佈局按鍵仍然用 `input_mode_pair`，兩者不衝突。
    public static let defaultItems: [StatusItem] = [
        StatusItem(source: .schemaName, tap: KeyAction(.schemaPicker, raw: "schema:picker")),
        StatusItem(source: .inputMode, tap: KeyAction(.inputModeToggle, raw: "input_mode:toggle")),
        StatusItem(source: .variant, tap: KeyAction(.toggleOption, ["simplification"],
                                                    raw: "toggle:simplification")),
        StatusItem(source: .page),
    ]
}

public struct Caret: Sendable {
    public var show = true
    public var color = RGBA.hex(0x3060C0FF)
    public var width = 1.5
    public var blink = true
}

public struct SelectionStyle: Sendable {
    public var color = RGBA.hex(0x3060C040)
    public var textColor = RGBA.hex(0x404040FF)
}

public struct Preedit: Sendable {
    public var show = true
    public var size = 16.0
    public var color = RGBA.hex(0x404040FF)
    public var background = RGBA.transparent
    public var paddingH = 6.0
    public var paddingV = 6.0
    public var cornerRadius = 0.0
    public var caret = Caret()
    public var selection = SelectionStyle()
}

/// §8.13（本輪新增）無障礙朗讀。
public struct AccessibilitySpec: Sendable {
    public var announceCandidates = AnnounceLevel.full
    public var candidateAnnouncement = "{label} {text} {comment}"
    public var announceInputMode = true
    public var announcePage = true
}

public struct Motion: Sendable {
    public var enabled = true
    public var respectReduceMotion = true
    public var candidateChangeMs = 90
    public var windowShowMs = 90
}

public struct Theme: Sendable {
    public var id: String
    public var revision = 1
    public var name = LocalizedString.empty
    public var description = LocalizedString.empty
    public var author = ""
    public var license = ""
    public var appearance = Appearance.light
    public var counterpart: String?
    public var ancestry: [String] = []
    public var palette: [String: RGBA] = [:]
    public var typography = Typography()
    public var metrics = Metrics()
    public var candidates = CandidateStyle()
    public var window = CandidateWindow()
    /// §8.6.6.3。桌面端解析但**不渲染** —— 見 `SyllableBar` 的註解。
    public var syllables = SyllableBar()
    public var statusBar = StatusBar()
    public var preedit = Preedit()
    public var accessibility = AccessibilitySpec()
    public var motion = Motion()

    public init(id: String) { self.id = id }
}
